### PR TITLE
Use sampling to reduce Search stat collection

### PR DESCRIPTION
**Pull request by brandon-m-skinner**
## Description

Presently we're sending too many stats too quickly. We'll start with a simple sampling and adjust as needed.

The main Search class and the Queue class got their own basic implementation of sampling so the tests could be adjusted to work properly without rewiring everything.

## Checklist

Please make sure the items below have been covered before requesting a review:

- [x] This change works and has been tested locally (or has an appropriate fallback).
- [x] This change works and has been tested on a Go sandbox.
- [x] This change has relevant unit tests (if applicable).
- [x] This change has relevant documentation additions / updates (if applicable).

## Steps to Test

1. Perform 10 search actions.
2. Every search action should record stats.
3. Apply PR.
4. Perform 10 search actions.
5. Only roughly half of the actions should record stats.

### CODE REVIEW

File: tests/search/includes/classes/test-class-queue.php [hunk 1]
```diff
@@ -1064,6 +1065,92 @@ public function test__delete_jobs_for_index_version() {
...
+		return [
+			[ array() ],
+			[ null ],
+			[ new \stdClass() ],
+			[ 5 ],
+			[ 8.6 ],
+		];
+	}
+
+	public function stat_sampling_invalid_value_param_data() {
```
Comment (pschoffer) [inline tests/search/includes/classes/test-class-queue.php:1078]: What do you think about moving the data providers right above the tests using it - e.g. - `test__maybe_update_stat_sampling_invalid_value_param` in this case.

It is a bit nitpicking though so feel free to ignore.

File: search/includes/classes/class-queue.php [hunk 2]
```diff
@@ -853,4 +855,28 @@ public function handle_index_limiting_start_timestamp() {
...
 	}
+
+	/**
+	 * Apply sampling to stats that are directly updated to keep stat sending in check.
+	 *
+	 * @param $stat string The stat to be possibly updated.
+	 * @param $value int The value to possibly update the stat with.
+	 */
+	public function maybe_update_stat( $stat, $value ) {
+		if ( ! is_string( $stat ) ) {
```
Comment (pschoffer) [inline search/includes/classes/class-queue.php:866]: Nice deffensive coding!

File: search/includes/classes/class-queue.php [hunk 3]
```diff
@@ -853,4 +855,28 @@ public function handle_index_limiting_start_timestamp() {
...
+			return;
+		}
+
+		if ( self::$stat_sampling_drop_value <= rand( 1, 10 ) ) {
+			return;
+		}
+
+		$value = intval( $value );
+
+		$this->statsd->update_stats( $stat, $value, 1, 'c' );
```
Comment (pschoffer) [inline search/includes/classes/class-queue.php:880]: Great catch with increment only updating by 1!

**Review by pschoffer — APPROVED** [review]
Looks great. I added minor note in tests, but leaving up to you if you want to address it or not.

**Comment by brandon-m-skinner** [comment]
r1286-stacks

**Pull request merged**